include/libcyphal/transport/udp/delegate.hpp

```diff
@@ -69,7 +75,150 @@ struct AnyUdpardTxMetadata
+            {
+                return 0;
+            }
+
+            // Find first fragment to start from (according to source `offset_bytes`).
+            //
+            std::size_t                  src_offset = 0;
+            const struct UdpardFragment* frag       = &payload_;
+            while ((nullptr != frag) && (offset_bytes >= (src_offset + frag->view.size)))
+            {
+                src_offset += frag->view.size;
+                frag = frag->next;
+            }
+
+            std::size_t dst_offset         = 0;
+            std::size_t total_bytes_copied = 0;
+
+            CETL_DEBUG_ASSERT(offset_bytes >= src_offset, "");
+            std::size_t view_offset = offset_bytes - src_offset;
+
+            while ((nullptr != frag) && (dst_offset < length_bytes))
+            {
+                CETL_DEBUG_ASSERT(nullptr != frag->view.data, "");
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+                FragSpan frag_span{static_cast<const cetl::byte*>(frag->view.data) + view_offset,  // NOSONAR cpp:S5356
+                                   std::min(frag->view.size - view_offset, length_bytes - dst_offset)};
+                CETL_DEBUG_ASSERT(frag_span.size() <= (frag->view.size - view_offset), "");
+
+                // Next nolint is unavoidable: we need offset from the beginning of the buffer.
+                // No Sonar `cpp:S5356` b/c we integrate here with libcanard raw C buffers.
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+                (void) std::memmove(destination + dst_offset, frag_span.data(), frag_span.size());  // NOSONAR cpp:S5356
+
+                src_offset += frag_span.size();
+                dst_offset += frag_span.size();
+                total_bytes_copied += frag_span.size();
+                CETL_DEBUG_ASSERT(dst_offset <= length_bytes, "");
+                frag        = frag->next;
+                view_offset = 0;
+            }
+
+            return total_bytes_copied;
+        }
+
+    private:
+        // MARK: Data members:
+
+        TransportDelegate& delegate_;
+        std::size_t        payload_size_;
+        UdpardFragment     payload_;
+
+    };  // UdpardMemory
+
+    struct SessionEvent
+    {
+        struct Message
+        {
+            struct Destroyed
+            {
+                PortId subject_id;
+            };
+
+            using Variant = cetl::variant<Destroyed>;
+        };
+
+        struct Request
+        {
+            struct Destroyed
+            {
+                PortId service_id;
+            };
+
+            using Variant = cetl::variant<Destroyed>;
+        };
+
+        struct Response
+        {
+            struct Destroyed
+            {
+                PortId service_id;
+            };
+
+            using Variant = cetl::variant<Destroyed>;
+        };
+
+        using Variant = cetl::variant<Message::Variant, Request::Variant, Response::Variant>;
```

Comment:
Here we have variant of variant. My idea is to keep session events kinda grouped...
- first, by their "kind" type (message, svc request & response)
- second, by their action type (inside of its kind)

Rationale:
- kinda more structured (hierarchical) way IMHO than "flat" approach
- extending actions of one kind (I anticipate that Req/Res probably will have more actions soon) won't touch other kind
- 2 `std::visit`-s (see below in code) quite nicely "flattens" the hierarchy when it comes to an event handler functor.